designs/isolate-cluster-upgrades.md

```diff
@@ -0,0 +1,221 @@
+
+```go
+type ClusterStatus struct {
+	...
+	// observedGeneration represents the .metadata.generation the last time the
+	// cluster was successfully reconciled.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// childrenObservedGeneration represents the sum of all cluster's linked
+	// objects's .metadata.generation the last time the cluster was successfully reconciled.
+	ChildrenObservedGeneration int64 `json:"childrenObservedGeneration,omitempty"`
+}
+```
+
+### Version skew
+Although it's possible today to continue upgrading a management cluster without updating the workload cluster's `Bundles` (the versions of all components in a EKS-A cluster), we lack a maximum skew version between management and workload clusters.
+
+Since this proposal makes the isolation between management and workload clusters even bigger, we propose to add to our documentation a maximum skew version support of one minor version. This means the management cluster can only be one EKS-A minor version after the workload clusters.
+
+Given clusters are not explicitly tagged to a particular EKS-A version (but implicitly through their `bundlesRef`), we recommend not adding a validation yet that prevents more than one minor version bumps. We recommend to tackle this problem first: moving to a model where clusters have a direct reference to its EKS-A version (**its** EKS-A version is whichever version was used to populate its component versions).
+
+We encourage to try to identify changes that can break controller support for clusters in old versions and make it clear in our release notes. Based on experience, we anticipate than, even if it not officially supported, users will still try to perform skip upgrades. The more obvious we make to them the situation were this can fail, the better.
+
+### Risks and doors
+Although it involves API changes, and as any other API change it should be approached with caution, we consider the proposed solution a two way door:
+
+* It's an internal mechanism not exposed to users that we are free to change as we please.
+* Even if this document is lengthy, the solution should be quite simple to implement, both in length and scope (overall codebase footprint). This should minimize the effort required to change it for a different solution.
+
+We can document the new API fields as internal and subject to change their meaning in future releases.
+
+#### Drift detection
+The biggest disadvantage of this solution is losing the ability to detect drift: if a user, or any other entity with enough permissions, changes the CAPI objects directly (hence affecting the real state of the cluster) we won't be able to react to this until any of the EKS-A objects composing the cluster config are updated. Although the controller shouldn't have any problem reconciling this new state of the cluster once the EKS-A cluster config is updated, this behavior is not ideal since it in a way breaks the expected Kubernetes control logic.
```

Comment:
First point: agreed, but this is already present today and I believe it's something it should be tackled independently from this doc. I think the proposed design is compatible with more configuration in thee API.


Second point: agreed, we should clearly document it. But we don't support updating capi objects, user shouldn't care about the behavior if they do that, because they shouldn't do it in the first place. And if they do, there are no guarantees.